README
Games-Sudoku-General is Copyright (C) 2005, 2006, 2008, 2011 by Thomas R.
Wyant, III

DESCRIPTION

Games::Sudoku::General is a module to solve the class of puzzle that
involves allocating symbols among sets such that no set contains more
than one of any symbol. Sudoku is the most common name for such
puzzles, though Number Place and Wasabi have also been used.

Each puzzle is considered to be made up of a number of cells, each of
which is a member of one or more sets, and each of which may contain
exactly one symbol. The contents of a number of cells are gven, and the
problem is to deduce the contents of the rest of the cells.

INSTALLATION

Most Perl users will want to install using their favorite of either
'cpan' or 'cpanp'. For either of these, installing
Games::Sudoku::General gets you everything. ActivePerl users will want
to use ActiveState's 'ppi', and install Games-Sudoku-General.

If for some reason neither of these works for you, you can download the
distribution and expand it using something like

    tar -xzf Games-Sudoku-General-9.999.tar.gz

Users of MSWin32 (and perhaps other operating systems) may need to
convert the line endings in the expanded files from Unix to their native
system before proceeding.

Then, issue one of the following two incantations:

    perl Makefile.PL
    make
    make test
    sudo make install

or

    perl Build.PL
    ./Build
    ./Build test
    sudo ./Build install

You should substitute the appropriate program name for 'make', eg nmake
(typically) under MSWin32, or mms or mmk under VMS. See
ftp://ftp.microsoft.com/Softlib/MSLFILES/nmake15.exe for a copy of
nmake for MSWin32 if you need it.

Unix users may want to use 'su' rather than 'sudo' for the install.
Non-unix users may need to be logged in to an administrative or
otherwise privileged accout, and turn on privileges (if applicable) for
the install.

Of course, since it's pure Perl, you can just expand the kit and drop
the .pm files into the Games/Sudoku directory (creating it if necessary)
in the appropriate place in your @INC directories.

OS-specific notes exist for (at least) MacOS (meaning, Mac OS 9
and lower), MSWin32, and VMS. See the relevant README files for
details.

LICENSING INFORMATION

This package is free software; you can redistribute it and/or modify it
under the same terms as Perl 5.10.0. For more details, see the full text
of the licenses in the directory LICENSES.

This program is distributed in the hope that it will be useful, but
without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.


Changes
0.015		2014-11-16	T. R. Wyant
   As of 2015-07-01, Makefile.PL and Build.PL will no longer prompt the
     user about installing sudokug, but default to installing it. The -n
     option will still suppress the installation. Also, the OS-specific
     functionality in these files will be removed.

0.014		2013-12-03	T. R. Wyant
   No code changes - just test tweaks.

0.013		2012-01-04	T. R. Wyant
   Remove unwanted cruft from Changes file, and bring it into compliance
      with Test::CPAN::Changes. Add xt/author/changes.t to be sure it
      stays that way.

0.012		2011-12-20	T. R. Wyant
   Update copy of Gnu GPL in LICENSES/Copying to have current FSF
      address.

0.011		2011-11-10	T. R. Wyant
   No changes from 0.010_90, except for version.

0.010_90	2011-11-04	T. R. Wyant
   'use 5.006002', since that is the oldest Perl I can build.
   Eliminate calls to UNIVERSAL::can as a function.
   Load optional module Win32::Clipboard by file name, mostly so code
      analyzers do not see it as an undeclared dependency.
   Require Test::More for testing.
   Add tests for POD and Perl::Critic, and for file presence and
      executability.
   Author testing is now by the authortest target, which requires
      building via Build.PL. Author testing is no longer supported when
      building via Makefile.PL.
   Moved the modification log from the module POD to the Changes file.
   Revise copyright and license statement.

0.010		2008-09-14	T. R. Wyant
   Fixed Build.PL to heed -y and -n options.

0.009		2008-07-21	T. R. Wyant
   'use 5.006', for 'our' at the very least.
   Update 'SEE ALSO'
   add =head1 LICENSE to POD.

0.008		2008-05-09	T. R. Wyant
   Tweak docs.
   Support unused cells.
   Add drop_set() method to undo add_set().
   Add 'null' attribute to generate a puzzle with no topology.
   Add 'quincunx' attribute to generate a quincunx (a.k.a.
     'Samurai Sudoku')

0.007		2006-09-30	T. R. Wyant
   Corrected example in topology attribute documentation,
     and other documentation tweaks.
   Moved General.pm to lib/Games/Sudoku.
   Added Build.PL
   In bin/sudokug, added the 'webcmd' pseudo-attribute and fixed the
     'source' command.

0.006		2006-07-03	T. R. Wyant
   Fixed problem with 'set corresponding'. Thanks
       to David Jelinek of Central Michigan University
       for diagnosing the problem and finding a
       solution.
   Corrected spelling.
   Eliminated Scalar::Util::looks_like_number, since
       apparently ActivePerl does not have it.
   Add copy() method and autocopy attribute, for getting
       generated puzzles onto the clipboard.
   Add paste() method, for loading puzzles from the
       clipboard.
   Add unload() method.
   In bin/sudokug, allowed input to be redirected or piped in, and added
       -filter option.

0.005		2005-12-20	T. R. Wyant
   Added generate() method and generation_limit
       attribute.
   Added rows attribute. This changes the default
       output for 'multi-faced' puzzles.

0.004		2005-12-16	T. R. Wyant
   Added Dion cube (via 'set cube number').

0.003		2005-12-13	T. R. Wyant
   Added 'set corresponding' and 'set max_tuple'.
   Added cubic sudoku (via 'set cube').
   Fixed horrendous inefficiency in backtrack logic.

0.002		2005-12-11	T. R. Wyant
   Format solution nicely for multi-character symbols.
   Fixed error in values eliminated by a hidden tuple.
   Recoded 'set sudokug' in terms of 'set brick', thus
     fixing an error in generating the small squares.
   Added method add_set(), and recoded 'set sudokux' in
     terms of this and 'set sudokug', thus fixing the
     same error that 'set sudoku' had.
   Put spaces in the result of scalar constraints_used.
   Spiffed up the POD.

0.001		2005-12-09	T. R. Wyant
   Initial release to CPAN.

# ex: set textwidth=72 :


t/TestDriver.pm

use strict;
use warnings;

package t::TestDriver;

use Games::Sudoku::General;
use Test;

our $VERSION = '0.001';

my $method;
my $number;
my $pass;
my $su;
my @todo;
my %variable;

sub execute {
my $self = shift;
my $handle = shift;

my $loc = tell $handle;

new ();

for ($pass = 0; $pass < 2; $pass++) {
    $number = 0;
    %variable = ();
    @todo = ();
    seek ($handle, $loc, 0);
    while (<$handle>) {
	s/^\s+//;
	next unless $_;
	next if m/^#/;
	s/\s+$//;
	my $outside = 1;
	my @args = map {
	    $outside++ % 2 ? (map {
		m/^<<(.+)/ ? do {
		    my $t = $1 . "\n"; my $r = '';
		    while (<$handle>) {last if $_ eq $t; $r .= $_}
		    $r } :
		m/^\$(.+)/ ? $variable{$1} || $1 :
		$_
		} split ('\s+', $_)) :
	    $_} split "'", $_;
	$variable{trace} and
	    print "#> ", join (' ', map {m/\s/ ? "'$_'" : $_} @args), "\n";
	my $verb = shift @args;
	if ($verb =~ m/\W/ || $verb =~ m/^_/) {
	    warn <<eod;
Error - Verb '$verb' is illegal.
eod
	    }
	  elsif (__PACKAGE__->can ($verb)) {
	    __PACKAGE__->$verb (@args);
	    }
	  elsif ($su->can ($verb)) {
	    next unless $pass;
	    $method = $verb;
	    $variable{result} = eval {$su->$verb (@args)} || $@ || '';
	    warn $@ if $@;
	    }
	  else {
	    warn <<eod;
Error - Verb '$verb' is unknown.
eod
	    }
	}


    unless ($pass) {plan tests => $number, todo => \@todo; }

    }
}

#	echo prints its arguments.

sub echo {
return unless $pass;
shift;
foreach (@_) {
    chomp;
    foreach (split '\n', $_) {
	s/^\.//;
	print "#         ", ' ' x length ($number), "$_\n" if $_
	}
    }
}

#	fowler translates Glenn Fowler's cell numbers into mine.

sub fowler {
$variable{result} = $_[1];
$variable{result} =~
    s/\[(\d+),(\d+)\]/'[' . (($1 - 1) * 9 + $2 - 1) . ']'/gem;
}

#	new instantiates a new Games::Sudoku::General object.

sub new {shift; $su = Games::Sudoku::General->new (@_)}

#	test compares the previous result to its first argument. It
#	also prints some front matter for the test, with the second
#	argument being the title (defaulting to the name of the
#	previous method), and subsequent arguments being echoed.

sub test {
shift;
$number++;
return unless $pass;
my $expect = shift;
chomp $expect;
my $title = shift || $method;
chomp $title;
print <<eod;
#
# Test $number - $title
eod
__PACKAGE__->echo (@_) if @_;
my $result = $variable{result};
chomp $result;
if ($result =~ m/\n/ || $expect =~ m/\n/) {
    foreach ([Expect => $expect], [Got => $result]) {
	printf "# %9s:\n", $_->[0];
	foreach (split '\n', $_->[1]) {print "#          $_\n"}
	}
    }
  else {
    print <<eod;
#    Expect: $expect
#       Got: $result
eod
    }
ok ($expect eq $result);
}

#	todo marks the previous test as a 'todo' item.

sub todo {push @todo, $number}

sub unload {$variable{result} = $su->_unload () if $pass}

#	var defines a pseudo-lexical variable, which can be substituted
#	into the command by prefixing the variable name with a '$'.
#	Note that $result is implicitly set by any
#	Games::Sudoku::General method.

sub var {shift; my $name = shift; $variable{$name} = "@_"}

1;


Build.PL
use 5.006002;	# For 'our', at the very least.

use strict;
use warnings;

use lib qw{ inc };
use base qw{ My::Module::Build };

use Config;
use FileHandle;
use Module::Build;
use My::Module::Meta;

(my $mbv = Module::Build->VERSION) =~ s/_//g;

my $meta = My::Module::Meta->new();

my %attr = (
    dist_author => 'Tom Wyant (wyant at cpan dot org)',
    dist_abstract => 'Solve Sudoku and related puzzles',
    module_name => 'Games::Sudoku::General',
    requires => $meta->requires(
	perl	=> $meta->requires_perl(),
    ),
    build_requires	=> $meta->build_requires(),
    configure_requires	=> {},	# Don't require Module::Build.
    get_options => {y => {}, n => {}},
    dynamic_config => 1,
    license => 'perl',
);

$mbv >= 0.28 and $attr{meta_merge} = {
    no_index => {
	directory => [qw{ inc t xt }],
    },
    resources => {
	bugtracker => 'https://rt.cpan.org/Public/Dist/Display.html?Name=Games-Sudoku-General',
    },
};

# Don't require Module::Build if we're making a distribution, since the
# the user may simply prefer ExtUtils::MakeMaker.

$mbv >= 0.34
    and $attr{auto_configure_requires} = !$meta->distribution();

#	Generate a builder object.

my $bldr = My::Module::Build->new (%attr);

#	Find out what the user wants to do.

my %opt = $bldr->args ();

my @exe_files;
my @clean_files;

my @possible_exes = ('sudokug');
print <<'EOD';

In addition to Games::Sudoku::General, you will get the following
executable:

  sudokug -- an interactive interface to Games::Sudoku::General.

If you do not want this, run Build.PL with the -n option.

Unless you are running MSWin32 or VMS (which are special-cased),
I will assume you're running some version of U*ix, and behave
accordingly.

>>>> NOTICE <<<<

Beginning with the first release after July 1 2015, sudokug will be
installed by default. If you do not want it installed, provide the -n
option when you run Build.PL.

I also intend to get rid of the OS-specific ad-hocery at this point.

EOD


if ($opt{n}) {
    print "Because you have asserted -n, the executables will not be installed.\n\n";
    }
  elsif ($opt{y}) {
    print "Because you have asserted -y, the executables will be installed.\n\n";
    @exe_files = @possible_exes;
    }
  else {
    foreach (@possible_exes) {
	push @exe_files, $_
	if $bldr->prompt ("Do you want to install $_?", 'n') =~ m/^y/i
	    ;
	}
    }

if (@exe_files) {
    if ($^O eq 'MSWin32') {
	@exe_files = map {"bin/$_"} @exe_files;
	foreach (@exe_files) {`pl2bat $_`}
	@clean_files = @exe_files =
	    grep {-e $_} map {"$_.bat"} @exe_files;
	}
      elsif ($^O eq 'VMS') {
	foreach my $fni (map {"[.bin]$_"} @exe_files) {
	    my $fno = "$fni.com";
	    my $fhi = FileHandle->new ("<$fni") or die <<eod;
Error - Unable to open $fni
        $!
eod
	    my $fho = FileHandle->new (">$fno") or die <<eod;
Error - Unable to open $fno
        $!
eod
	    print $fho "$Config{startperl}\n";
	    while (<$fhi>) {print $fho $_}
	    }
	@clean_files = @exe_files = map {"[.bin]$_.com"} @exe_files;
	}
      else {
	@exe_files = map {"bin/$_"} @exe_files;
	}
    }

##my $vers = $] >= 5.008 ? '-5.8' : '';

#	Tell Module::Build what we want to do.

$bldr->add_to_cleanup (\@clean_files);
$bldr->script_files (\@exe_files);

#	Generate the build script, at long last.

$bldr->create_build_script ();

# ex: set textwidth=72 :


inc/My/Module/Meta.pm
package My::Module::Meta;

use 5.006002;

use strict;
use warnings;

use Carp;

sub new {
    my ( $class ) = @_;
    ref $class and $class = ref $class;
    my $self = {
	distribution => $ENV{MAKING_MODULE_DISTRIBUTION},
    };
    bless $self, $class;
    return $self;
}

sub build_requires {
    return +{
##	'Test::More'	=> 0.40,
	'Test::More'	=> 0.88,	# Because of done_testing().
    };
}

sub distribution {
    my ( $self ) = @_;
    return $self->{distribution};
}

sub requires {
    my ( $self, @extra ) = @_;

    return {
	'Carp'			=> 0,
	'constant'		=> 0,
	'strict'		=> 0,
	'warnings'		=> 0,
	@extra,
    };
}

sub requires_perl {
    return 5.006002;
}


1;

__END__

=head1 NAME

My::Module::Meta - Information needed to build Astro::SpaceTrack

=head1 SYNOPSIS

 use lib qw{ inc };
 use My::Module::Meta;
 my $meta = My::Module::Meta->new();
 use YAML;
 print "Required modules:\n", Dump(
     $meta->requires() );

=head1 DETAILS

This module centralizes information needed to build C<App::Satpass2>. It
is private to the C<App::Satpass2> package, and may be changed or
retracted without notice.

=head1 METHODS

This class supports the following public methods:

=head2 new

 use lib qw{ inc };
 my $meta = My::Module::Meta->new();

This method instantiates the class.

=head2 build_requires

 use YAML;
 print Dump( $meta->build_requires() );

This method computes and returns a reference to a hash describing the
modules required to build the C<Astro::Coord::ECI> package, suitable for
use in a F<Build.PL> C<build_requires> key, or a F<Makefile.PL>
C<< {META_MERGE}->{build_requires} >> key.

=head2 distribution

 if ( $meta->distribution() ) {
     print "Making distribution\n";
 } else {
     print "Not making distribution\n";
 }

This method returns the value of the environment variable
C<MAKING_MODULE_DISTRIBUTION> at the time the object was instantiated.

=head2 requires

 use YAML;
 print Dump( $meta->requires() );

This method computes and returns a reference to a hash describing
the modules required to run the C<App::Satpass2> package, suitable for
use in a F<Build.PL> C<requires> key, or a F<Makefile.PL> C<PREREQ_PM>
key. Any additional arguments will be appended to the generated hash. In
addition, unless L<distribution()|/distribution> is true,
configuration-specific modules may be added.

=head2 requires_perl

 print 'This package requires Perl ', $meta->requires_perl(), "\n";

This method returns the version of Perl required by the package.

=head1 ATTRIBUTES

This class has no public attributes.


=head1 ENVIRONMENT

=head2 MAKING_MODULE_DISTRIBUTION

This environment variable should be set to a true value if you are
making a distribution. This ensures that no configuration-specific
information makes it into F<META.yml>.


=head1 SUPPORT

Support is by the author. Please file bug reports at
L<http://rt.cpan.org>, or in electronic mail to the author.

=head1 AUTHOR

Thomas R. Wyant, III F<wyant at cpan dot org>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010-2014 by Thomas R. Wyant, III

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl 5.10.0. For more details, see the full text
of the licenses in the directory LICENSES.

This program is distributed in the hope that it will be useful, but
without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.

=cut

# ex: set textwidth=72 :


Makefile.PL
use 5.006002;	# For 'our', at the very least.

use strict;
use warnings;

use lib qw{ inc };

use Config;
use ExtUtils::MakeMaker qw{WriteMakefile prompt};
use FileHandle;
use Getopt::Std;
use My::Module::Meta;

my $meta = My::Module::Meta->new();

my %opt;
getopts ('ny', \%opt) or die <<eod;
The only legal options are
  -n = answer all questions 'no';
  -y = answer all questions 'yes'.
eod

my @exe_files;
my @clean_files;

print <<'EOD';

In addition to Games::Sudoku::General, you will get the following
executable:

  sudokug -- an interactive interface to Games::Sudoku::General.

Unless you are running MSWin32 or VMS (which are special-cased),
I will assume you are running some version of U*ix, and behave
accordingly.

If you do not want this executable installed, rerun this script
with the -n qualifier.

>>>> NOTICE <<<<

Beginning with the first release after July 1 2015, sudokug will be
installed by default. If you do not want it installed, provide the -n
option when you run Makefile.PL.

I also intend to get rid of the OS-specific ad-hocery at this point.

EOD

if ($opt{n}) {
    print "Because you have asserted -n, the executable will not be installed.\n\n";
    }
  elsif ( $opt{y} ) {
    push @exe_files, qw{ sudokug };
    print "Because you have asserted -y, the executable will be installed.\n";
    }
  else {
    foreach (qw{sudokug}) {
	push @exe_files, $_
	    if prompt ("Do you want to install $_?", 'n') =~ m/^y/i;
	}
    }

if (@exe_files) {
    if ($^O eq 'MSWin32') {
	@exe_files = map {"bin/$_"} @exe_files;
	foreach (@exe_files) {`pl2bat $_`}
	@clean_files = @exe_files =
	    grep {-e $_} map {"$_.bat"} @exe_files;
	}
      elsif ($^O eq 'VMS') {
	foreach my $fni (map {"[.bin]$_"} @exe_files) {
	    my $fno = "$fni.com";
	    my $fhi = FileHandle->new ("<$fni") or die <<eod;
Error - Unable to open $fni
        $!
eod
	    my $fho = FileHandle->new (">$fno") or die <<eod;
Error - Unable to open $fno
        $!
eod
	    print $fho "$Config{startperl}\n";
	    while (<$fhi>) {print $fho $_}
	    }
	@clean_files = @exe_files = map {"[.bin]$_.com"} @exe_files;
	}
      else {
	@exe_files = map {"bin/$_"} @exe_files;
	}
    }


##my $vers = $] >= 5.008 ? '-5.8' : '';
my $vers = '';

(my $mmv = ExtUtils::MakeMaker->VERSION) =~ s/_//g;
my %attr = (
##  (MM->can ('signature_target') ? (SIGN => 1) : ()),
    NAME => 'Games::Sudoku::General',
    VERSION_FROM => 'lib/Games/Sudoku/General.pm',
    PREREQ_PM => $meta->requires(),
    PL_FILES => {},	# Prevent old MakeMaker from running Build.PL
    EXE_FILES => \@exe_files,
#   'linkext' => {LINKTYPE => ''},
    'dist' => {COMPRESS => 'gzip', SUFFIX => 'gz'},
    realclean => {FILES => join ' ', @clean_files},
    AUTHOR => 'Tom Wyant (wyant at cpan dot org)',
    ABSTRACT => 'Solve Sudoku and related puzzles',
    BINARY_LOCATION => "$Config{archname}$vers/Games-Sudoku-General.tar.gz",
);


$mmv >= 6.31 and $attr{LICENSE} = 'perl';

if ( $mmv >= 6.4501 ) {
    $attr{META_ADD} = {
	no_index => {
	    directory => [ qw{ inc t xt } ],
	},
    };    
    $attr{META_MERGE} = {
	resources => {
	    bugtracker => 'https://rt.cpan.org/Public/Dist/Display.html?Name=Games-Sudoku-General',
	    license => 'http://dev.perl.org/licenses/',
	},
    };
}

$mmv >= 6.4701 and $attr{MIN_PERL_VERSION} = $meta->requires_perl();

$mmv >= 6.5501
    and $attr{BUILD_REQUIRES} = $meta->build_requires()
    or $mmv >= 6.4501
    and $attr{META_MERGE}{build_requires} = $meta->build_requires();

WriteMakefile(%attr);

# ex: set textwidth=72 :
